src/ddscxx/include/dds/sub/detail/SamplesHolder.hpp

```diff
@@ -139,8 +139,39 @@ class CDRSamplesHolder : public SamplesHolder
       uint32_t cpp_sample_size = this->samples_.delegate()->length();
       for (uint32_t i = 0; i < cpp_sample_size; ++i)
       {
+        struct ddsi_serdata * current_blob = cdr_blobs[i];
+#ifdef DDS_HAS_SHM
+        // if the data is over iox_chunk
+        if (current_blob->iox_chunk && current_blob->iox_subscriber) {
+          // get the data
+          org::eclipse::cyclonedds::topic::CDRBlob & sample_data =
+            (*this->samples_.delegate())[i].delegate().data();
+          // update the header
+          memcpy(sample_data.encoding().data(), current_blob->iox_chunk, 4);
+          // update the data kind
+          sample_data.kind(
+            static_cast<org::eclipse::cyclonedds::topic::BlobKind>(current_blob->kind));
+          if (sample_data.kind() != org::eclipse::cyclonedds::topic::BlobKind::Empty) {
+            // get the size
+            auto iox_header = iceoryx_header_from_chunk(current_blob->iox_chunk);
+            // if the iox chunk has the data in serialized form
+            if (iox_header->shm_data_state == IOX_CHUNK_CONTAINS_SERIALIZED_DATA) {
+              // get the actual data
+              sample_data.payload().assign(
+                reinterpret_cast<uint8_t *>(current_blob->iox_chunk) + 4,
+                reinterpret_cast<uint8_t *>(current_blob->iox_chunk) + iox_header->data_size);
+            } else {
+              // this shouldn't never happen
+              ISOCPP_THROW_EXCEPTION(ISOCPP_PRECONDITION_NOT_MET_ERROR,
+                "The received sample over SHM is not in serialized form");
```

Comment:
Probably this assumption is not always true!

Reply:
The scenario that the iox chunk containing the raw data should also be handled here, I will modify this accordingly

Reply:
It's probably worth splitting this out into a separate function. Also we should do something about the `+4`s ...

Reply:
Right., the 4 should be the CDR header size constant that should exist somewhere, if not has be created. 

@sumanth-nirmal I am not sure we can get raw (original, unserialized) data here, but if so yes it should be handled This case must have been handled before (fixed-size types) somewhere though ...